### DISCUSSION TITLE
Arrow integration in C++?

### DISCUSSION THREAD

**Discussion by ghost** [duckdb/duckdb#2776] [Q&A]
Hi,

I am trying to do Arrow DuckDB zero-copy integration in C++. It seems currently this is only supported from a scripting language like Python or R? Are there plans to expose this in the C++ public API?

I read the R code and tried to do the same thing in C++, but ran into an issue where I need to create a ``ReplacementScan`` for Arrow, which seems to require the user to provide DuckDB with a user function with signature ``std::unique_ptr<duckdb::TableFunctionRef>(const std::string& table_name, void* data)``  But ``duckdb::TableFunctionRef`` is not part of the public API in the amalgamation build, so I hit a brick wall there. 

Thanks

**Comment by ghost** [comment]
Well I got it mostly working. I realize that that this is a work-in-progress, and I'm poking around in the internals of the library, but some thoughts I had: I had to translate the DuckDB expression AST to Arrow expression AST in C++ myself, in order to get projection and filtering working. I could not find this implemented anywhere in the C++ DuckDB or Arrow library, this "middleware" logic is instead found in the Python and R language bindings for Arrow and DuckDB. I assume this was done in order to maintain independence between Arrow and DuckDB Python/R packages. The C++ code that I wrote has direct dependencies on both C++ libraries, so it's difficult to see how it could be part of either library.

**Reply by hannes** [reply]
Indeed that was the reasoning there.

**Comment by readwriteexecute777** [comment]
Hello! Is there any update on this? I would like to be able to query in-memory Arrow objects using the C or C++ APIs. Thanks!

**Comment by wangfenjin** [comment]
Maybe you can try this https://github.com/datafusion-contrib/datafusion-c
